README.md
# Bioi
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/Bioi)](https://cran.r-project.org/package=Bioi)
[![Travis-CI Build Status](https://travis-ci.org/zcolburn/Bioi.svg?branch=master)](https://travis-ci.org/zcolburn/Bioi)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/zcolburn/Bioi?branch=master&svg=true)](https://ci.appveyor.com/project/zcolburn/Bioi)
[![codecov](https://codecov.io/gh/zcolburn/Bioi/branch/master/graph/badge.svg)](https://codecov.io/gh/zcolburn/Bioi)


## Overview
Bioi is an R package containing implementations of solutions to common cell biology image processing problems. In particular, Bioi provides functions to perform connected component labeling on 1, 2, or 3 dimensional arrays, single linkage clustering on n-dimensional arrays, and identification of the points in one data set that are closest to each point in a second data set.


## Installation
The Bioi project can be found on its [GitHub repository](https://github.com/zcolburn/Bioi). It can be installed from that source using functionality available in the devtools package.


```r
install.packages("devtools")
library(devtools)
install_github("zcolburn/Bioi")
library(Bioi)
```


## Key functionality
### N-dimensional single linkage clustering
The objective of single linkage clustering is to place all points into groups such that all points within a group can be reached from any other point in the group by crossing bridges between points that are less than a critical separation distance. A small critical separation distance may result in a larger number of groups being identified. In contrast, a large critical separation distance may result in fewer groups being identified.


Using the function `euclidean_linker`, single linkage clustering can be performed in 1 or more dimensions. The function works in three modes: unpartitioned, partitioned, and parallelized. For small sample sizes (number of points less than `partition_req`) the unpartitioned method is used. For larger sample sizes the partitioned method is used.


The partitioned method works by iteratively dividing the data into smaller and smaller subsections until each partition contains fewer points than `partition_req`. The unpartitioned method is then used on each subsection before combining the data from each subsection.


The parallelized method works similarly to the partitioned method but operates in parallel. The number of cores to use can be specified by `num_cores`. To prevent too many threads from being generated `parallel_call_depth` can be specified. A higher depth will result in more threads being generated.


### Connected component labeling
A common image processing task is to group all connected "object-positive" pixels in an image into single groups. The connected component labeling function implemented here can be used on 1, 2, or 3-dimensional arrays representing 1, 2, or 3-dimensional "images". This functionality can be acessed using the `find_blobs` function.


### Determine the nearest neighbor in an alternate set of points
Photoactivated localization microscopy (PALM) data results in large numbers of protein localizations being identified. A common task when working with dual channel PALM data is to identify the distance separating points in one data set from points in a second data set. The function `find_min_dists` identifies the nearest neighbor to a point in a second data set and its distance from the point of interest.




NEWS.md
# Bioi 0.2.9

* Fixed an invalid memory read issue in euclidean_linker_cpp.cpp.


* Fixed a namespace NOTE related to testthat  that occurred on fedora, solaris, and osx.


* Changed 'abs' to 'fabs' in euclidean_linker_cpp.cpp which was detected on debian with using the clang compiler.


tests/testthat/test.euclidean_linker.R
context("euclidean_linker")

test_that(
  "invalid inputs produce an error",
  {
    expect_error({
      input <- data.frame(x = 1:10, y = 1:10)
      crit_dist <- 1.5
      euclidean_linker(input, crit_dist)
    })

    expect_error({
      input <- as.matrix(data.frame(x = 1:10, y = 1:10))
      crit_dist <- "TRUE"
      euclidean_linker(input, crit_dist)
    })

    expect_error({
      input <- as.matrix(data.frame(x = 1:10, y = 1:10))
      crit_dist <- TRUE
      euclidean_linker(input, crit_dist)
    })
  }
)


test_that(
  "outputs are valid",
  {
    expect_true({
      vec_length <- 10000
      input <- as.matrix(data.frame(x = 1:vec_length, y = 1:vec_length))
      crit_dist <- 1.5
      result <- euclidean_linker(input, crit_dist)
      all(result == 1)
    })

    expect_equal({
      set.seed(10)
      input <- as.matrix(data.frame(x = rnorm(10), y = rnorm(10)))
      crit_dist <- 0.4
      euclidean_linker(input, crit_dist)
    },
    c(7, 6, 2, 4, 8, 9, 3, 5, 1, 6)
    )

    expect_equal({
      set.seed(10)
      num <- 1000
      input <- as.matrix(data.frame(
        x = runif(num)*40,
        y = runif(num)*40,
        z = runif(num)*40
      ))
      crit_dist <- 5
      result <- euclidean_linker(
        input,
        crit_dist,
        partition_req = 200,
        run_parallel = FALSE
      )
      unique(result)
    },
      c(3, 2, 1)
    )

    expect_equal({
      mat <- matrix(TRUE, nrow = 501, ncol = 501)
      result <- Bioi::find_blobs(mat)
      unique(as.vector(result))
    },
    1
    )
  }
)




R/find_blobs.R
#' @title Assign all neighboring pixels the same group number.
#'
#' @description
#' Perform connected-component labeling to group continuous, thresholded
#' objects in 3-dimensional arrays.
#'
#' This function takes a vector, matrix, or 3-dimensional array where each
#' element is TRUE if it corresponds to an object-positive index or FALSE if it
#' corresponds to a background index. An object of the same dimension as the
#' input is returned. All connected object indices take the value of their
#' group number and all background indices take the value NA.
#'
#' @param arr A vector, matrix, or 3-dimensional array where object-positive
#' elements are denoted by the value TRUE and background elements are denoted
#' by the value FALSE.
#' @param use_prog_bar TRUE/FALSE indicating whether a progress bar should be
#' used. This is only available when run_parallel is FALSE.
#' @param run_parallel TRUE/FALSE indicating whether operations should be
#' performed in parallel. This is only valid if partitioning is performed.
#' @param num_cores The number of cores to use if running in parallel.
#' @param partition_req The minimum number of points required to create a new
#' partition.
#' @param parallel_call_depth The number of levels of partitioning that should
#' be performed before terminating calls to run operations in parallel. The
#' number of threads opened when running in parallel is equal to
#' 2^(parallel_call_depth)*num_cores.
#'
#' @author Zach Colburn
#'
#' @examples
#' # Generate a random matrix.
#' set.seed(10)
#' mat <- matrix(runif(70), nrow = 7)
#'
#' # Arbitrarily say that everything below 0.8 is background.
#' logical_mat <- mat > 0.8
#'
#' # Find blobs.
#' find_blobs(logical_mat)
#'
#' @export
#'
#' @importFrom assertthat assert_that noNA is.number
find_blobs <- function(
  arr,
  use_prog_bar = TRUE,
  run_parallel = FALSE,
  num_cores = NULL,
  partition_req = NULL,
  parallel_call_depth = 3
) {
  # Perform type checking.
  assert_that(is.vector(arr) || is.matrix(arr) || is.array(arr))
  assert_that(length(arr) >= 1)
  assert_that(class(arr[1]) == "logical")
  assert_that(noNA(arr))
  if(is.null(partition_req)){partition_req <- 1000000}
  assert_that(
    is.number(partition_req) &&
      (as.integer(partition_req) == partition_req) &&
      (partition_req > 0)
  )

  # Get object class and attributes.
  initial_class <- class(arr)
  arr_attributes <- attributes(arr)

  # Convert arr to an array and store object indices in "input".
  arr <- as.array(arr)
  input <- as.matrix(which(arr, arr.ind = TRUE))

  # Initialize the output object.
  output <- array(NA, dim = dim(arr), dimnames = dimnames(arr))

  # If there are no object indices then return output without performing
  # any more operations.
  if(nrow(input) == 0){
    print("There are no objects to link!")
    return(output)
  }

  # Use the euclidean_linker_cpp function to link neighboring object indices.
  # Since the critical distance is sqrt(3), every neighboring object index
  # (both horizontally/vertically and diagonally) will be joined in 1, 2, or 3
  # dimensions. "links" is a vector of group numbers.
  links <- euclidean_linker(
    input,
    critDist = sqrt(3),
    use_prog_bar = use_prog_bar,
    run_parallel = run_parallel,
    num_cores = num_cores,
    partition_req = partition_req,
    parallel_call_depth = parallel_call_depth,
    find_blobs = TRUE
  )

  # Assign indices in the output object their respective group number.
  output[input] <- links

  # Convert the output object to the class of the original arr object.
  if(initial_class == "logical"){
    output <- as.vector(output)
  }else if(initial_class == "matrix"){
    output <- as.matrix(output)
  }

  # Restore the object's attributes. This is mainly for restoring vector
  # element names and matrix row/column names).
  attributes(output) <- arr_attributes

  # Return the output object.
  return(output)
}


#' @title Assign all neighboring pixels the same group number.
#'
#' @description
#' This function is deprecated. It now calls the more efficient find_blobs
#' method.
#'
#' This function takes a matrix corresponding to a thresholded image and
#' returns a matrix of the same size, where all adjacent, thesholded pixels
#' are the same integer corresponding to that object's cluster number.
#'
#' @param img A thresholded matrix (where non-object pixels are assigned a
#' value of 0).
#' @param pixRange This parameter is now obsolete. Previously, the parameter
#' denoted an integer number of pixels to specify a search region. Execution
#' was faster when this value was small. However, the value needed to be larger
#' than the diameter of the largest continuous object in the image.
#'
#' @author Zach Colburn
#'
#' @examples
#' # Generate a random matrix.
#' set.seed(10)
#' mat <- matrix(runif(70), nrow = 7)
#'
#' # Arbitrarily say that everything below 0.8 is background.
#' mat[mat < 0.8] <- 0
#'
#' # Find blobs.
#' identify_thresholded_objects(mat)
#'
#' @export
#'
#' @importFrom assertthat assert_that is.number
identify_thresholded_objects <- function(img, pixRange = 50){
  # Perform type checking. This function was meant to receive different inputs
  # than the function find_blobs which replaces it. Type checking is performed
  # to ensure object inputs are backwards compatible.
  assert_that(class(img) == "matrix")
  assert_that(length(img) >= 1)
  assert_that(class(img[1]) %in% c("integer", "numeric"))
  assert_that(nrow(img) >= 1)
  assert_that(ncol(img) > 0)

  # If pixRange has changed, then inform the user that its use is deprecated.
  assert_that(is.number(pixRange))
  if(pixRange != 50){
    print("The use of pixRange is deprecated.")
  }

  # Convert the input to a logical matrix.
  img <- img != 0

  # Perform connected component labeling.
  find_blobs(img)
}

R/RcppExports.R
# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @title Return the group number for each localization.
#'
#' @description
#' Group PALM/iPALM localizations based on their physical separation distance
#'
#' PALM/iPALM data results in a list of spatial coordinates for fluorophore
#' localizations. This function groups nearby localizations if they are within
#' the provided critical distance from each other.
#'
#' @param input A numeric matrix where each row is a localization and each
#' column is a spatial axis.
#' @param critDist The critical distance for which localizations nearer than
#' this distance are deemed part of the same group.
#' @param use_prog_bar A logical indicating whether a progress bar should be
#' used. This must be set to false when running in parallel.
#'
#' @author Zach Colburn
#'
#' @examples
#' # Function call
#' \dontrun{euclidean_linker_cpp(inputMatrix, critDist)}
#'
#' @import Rcpp
#'
#' @useDynLib Bioi, .registration = TRUE
.euclidean_linker_cpp <- function(input, critDist, use_prog_bar = TRUE) {
    .Call(`_Bioi_euclidean_linker_cpp`, input, critDist, use_prog_bar)
}

#' @title For all points in matrix 1, return the distance to and index of the
#' nearest point in matrix 2.
#'
#' @description
#' Find the shortest distance between each point in one data set and the points
#' in a second set.
#'
#' This function determines the distance between every point in data set 1 and
#' the points in data set 2. Unlike this function's naive counterpart,
#' find_min_dists, this function divides the PALM/iPALM localization data
#' into blocks, opeartes on the data in each block, and then performs linking
#' operations on neighboring blocks.
#'
#' @param mOne A numeric matrix where each row is a localization and each
#' column is a spatial axis.
#' @param mTwo A numeric matrix with the same number of columns as mOne.
#'
#' @author Zach Colburn
#'
#' @examples
#' set.seed(10)
#'
#' mOne <- as.matrix(data.frame(
#' x = rnorm(10),
#' y = rbinom(10, 100, 0.5),
#' z = runif(10)
#' ))
#'
#' mTwo <- as.matrix(data.frame(
#' x = rnorm(20),
#' y = rbinom(20, 100, 0.5),
#' z = runif(20)
#' ))
#'
#' find_min_dists_cpp(mOne, mTwo)
#'
#' @import Rcpp
#'
#' @useDynLib Bioi, .registration = TRUE
.find_min_dists_cpp <- function(mOne, mTwo) {
    .Call(`_Bioi_find_min_dists_cpp`, mOne, mTwo)
}



R/find_min_dists.R
#' @title For all points in matrix 1, return the distance to and index of the
#' nearest point in matrix 2.
#'
#' @description
#' Find the shortest distance between each point in one data set and the points
#' in a second set.
#'
#' This function determines the distance between every point in mOne and the
#' nearest point in mTwo.
#'
#' @param mOne A numeric matrix where each row is a localization and each
#' column is a spatial axis.
#' @param mTwo A numeric matrix with the same number of columns as mOne.
#'
#' @author Zach Colburn
#'
#' @examples
#' # Generate random data.
#' set.seed(10)
#'
#' mOne <- as.matrix(data.frame(
#' x = rnorm(10),
#' y = rbinom(10, 100, 0.5),
#' z = runif(10)
#' ))
#'
#' mTwo <- as.matrix(data.frame(
#' x = rnorm(20),
#' y = rbinom(20, 100, 0.5),
#' z = runif(20)
#' ))
#'
#' # Find the minimum distance between each point in mOne and the points in
#' # mTwo.
#' find_min_dists(mOne, mTwo)
#'
#' @export
#'
#' @importFrom assertthat assert_that
find_min_dists <- function(mOne, mTwo) {
  # Perform type checking.
  assert_that(class(mOne) == "matrix")
  assert_that(length(mOne) >= 1)
  assert_that(class(mOne[1]) %in% c("integer", "numeric"))
  assert_that(nrow(mOne) >= 1)
  assert_that(ncol(mOne) > 0)

  assert_that(class(mTwo) == "matrix")
  assert_that(length(mTwo) >= 1)
  assert_that(class(mTwo[1]) %in% c("integer", "numeric"))
  assert_that(nrow(mTwo) >= 1)
  assert_that(ncol(mTwo) > 0)

  # Find the minimum distance between each point in mOne and the points in mTwo.
  .find_min_dists_cpp(mOne, mTwo)
}


R/perform_grouping.R
#' Group PALM/iPALM localizations based on their physical separation distance
#'
#' PALM/iPALM data results in a list of spatial coordinates for fluorophore
#' localizations. This function groups nearby localizations if they are within
#' the provided critical distance from each other.
#'
#' @title Return the group number for each localization.
#'
#' @param input A numeric matrix where each row is a localization and each
#' column is a spatial axis.
#' @param critDist The critical distance for which localizations nearer than
#' this distance are deemed part of the same group.
#' @param use_prog_bar TRUE/FALSE indicating whether a progress bar should be
#' used. This is only available when run_parallel is FALSE.
#'
#' @author Zach Colburn
#'
#' @export
#'
#' @importFrom assertthat assert_that
#' @importFrom stats var
.perform_grouping <- function(
  input,
  critDist,
  use_prog_bar = TRUE
) {
  # Complete type checking was performed at the start of "euclidean_linker"
  # which calls this function. As such, only critical type checks are
  # performed. This is meant mainly to ensure that data passed from the
  # function ".perform_partitioning" which is called by "euclidean_linker" is
  # in the correct format.
  assert_that(class(input) == "matrix")
  assert_that(class(input[1]) %in% c("integer", "numeric"))
  assert_that(nrow(input) >= 1)

  # Return 1 if there is only a single input point.
  if(nrow(input) == 1){
    return(1)
  }

  # Reorder the columns such that the first column accounts for the most
  # variance in position and the last column accounts for the least variance in
  # position.
  column_variances <- apply(input, MARGIN = 2, FUN = var)
  new_column_order <- order(column_variances, decreasing = TRUE)
  input <- input[, new_column_order, drop = FALSE]

  # Order elements in input.
  ordered_elements <- do.call(order, as.data.frame(input))

  # Arrange the rows in input based on the ordering performed above.
  input <- input[ordered_elements,,drop = FALSE]

  # Identify groups.
  groups <- .euclidean_linker_cpp(
    input,
    critDist,
    use_prog_bar = use_prog_bar
  )

  # Change the order of groups in the output to undo the sorting performed
  # above.
  output <- vector(mode = "numeric", length = length(groups))
  output[ordered_elements] <- groups

  # Change group numbers such that they range from 1 to number of groups.
  output <- as.numeric(factor(output))

  # Return the output.
  output
}



R/perform_partitioning.R
#' Group PALM/iPALM localizations based on their physical separation distance
#'
#' PALM/iPALM data results in a list of spatial coordinates for fluorophore
#' localizations. This function groups nearby localizations if they are within
#' the provided critical distance from each other.
#'
#' @title Return the group number for each localization.
#'
#' @param input A numeric matrix where each row is a localization and each
#' column is a spatial axis.
#' @param critDist The critical distance for which localizations nearer than
#' this distance are deemed part of the same group.
#' @param use_prog_bar TRUE/FALSE indicating whether a progress bar should be
#' used. This is only available when run_parallel is FALSE.
#' @param run_parallel TRUE/FALSE indicating whether operations should be
#' performed in parallel. This is only valid if partitioning is performed.
#' @param num_cores The number of cores to use if running in parallel.
#' @param partition_req The minimum number of points required to create a new
#' partition.
#' @param parallel_call_depth The number of levels of partitioning that should
#' be performed before terminating calls to run operations in parallel. The
#' number of threads opened when running in parallel is equal to
#' 2^(parallel_call_depth)*num_cores.
#' @param min_gap The minimum width of any dimension created during paritioning.
#'
#' @author Zach Colburn
#'
#' @export
#'
#' @importFrom assertthat assert_that is.number is.flag
#' @importFrom parallel parLapply makeCluster detectCores
#' @importFrom igraph components graph.data.frame
#' @importFrom dplyr mutate_all
.perform_partitioning <- function(
  input,
  critDist,
  use_prog_bar = TRUE,
  run_parallel = FALSE,
  num_cores = NULL,
  partition_req = 5000,
  parallel_call_depth = 3,
  min_gap = NULL
) {
  # Perform essential type checking.
  assert_that((class(input) == "matrix"))
  assert_that(is.number(min_gap))

  # Define the ungrouped value. Don't change this.
  no_group <- -1

  # Check whether parallel_call_depth is 1. If so set run_parallel to FALSE.
  if(parallel_call_depth == 1){
    run_parallel <- FALSE
  }

  # Handle the case of only a single input point.
  if(nrow(input) == 1){
    return(1)
  }

  # Specify partitions.
  use_prog_bar <- FALSE
  ranges <- apply(input, 2, function(item){max(item) - min(item)})
  dim_to_split <- which.max(ranges)

  # If the partition is too small, then run euclidean linker on this data set
  # without performing partitioning, regardless of how many points are in the
  # partition. Also, if the number of points in this partition is less than the
  # partition requirement, then perform grouping.
  if(
    (min(ranges) <= min_gap) ||
    (nrow(input) <= partition_req)
  ){
    groups <- .perform_grouping(
      input,
      critDist,
      use_prog_bar = use_prog_bar
    )
    return(groups)
  }

  # Determine partitioning bounds.
  max_pos <- max(input[,dim_to_split])
  min_pos <- min(input[,dim_to_split])
  mid_pos <- min_pos + (max_pos - min_pos)/2

  # Assign points to partitions and stem group.
  partition <- rep(1, nrow(input))
  partition[input[,dim_to_split] > mid_pos] <- 2
  stem <- (input[,dim_to_split] >= (mid_pos - min_gap/2)) &
    (input[,dim_to_split] < (mid_pos + min_gap/2))

  # If the stem and both partitions don't all have points then group now.
  if(!(
    (sum(stem) >= 1) &&
    (sum(partition == 1) >= 1) &&
    (sum(partition == 2) >= 1)
  )){
    groups <- .perform_grouping(
      input,
      critDist,
      use_prog_bar = use_prog_bar
    )
    groups <- as.numeric(factor(groups))
    return(groups)
  }

  # Perform grouping.
  if(!run_parallel){
    groups <- lapply(
      list(
        input[partition == 1,, drop = FALSE],
        input[partition == 2,, drop = FALSE],
        input[stem,, drop = FALSE]
      ),
      function(item){
        .perform_partitioning(
          item,
          critDist,
          use_prog_bar = FALSE,
          run_parallel = run_parallel,
          num_cores = NULL,
          partition_req = partition_req,
          min_gap = min_gap
        )
      }
    )
  }else{
    # Perform operations in parallel.
    groups <- parLapply(
      makeCluster(num_cores),
      list(
        input[partition == 1,, drop = FALSE],
        input[partition == 2,, drop = FALSE],
        input[stem,, drop = FALSE]
      ),
      function(
        item,
        critDist,
        run_parallel,
        num_cores,
        partition_req,
        parallel_call_depth,
        .perform_partitioning,
        min_gap
      ){
        .perform_partitioning(
          item,
          critDist,
          use_prog_bar = FALSE,
          run_parallel = run_parallel,
          num_cores = num_cores,
          partition_req = partition_req,
          parallel_call_depth = parallel_call_depth - 1,
          min_gap = min_gap
        )
      }, critDist,
      run_parallel,
      num_cores,
      partition_req,
      parallel_call_depth,
      .perform_partitioning,
      min_gap
    )
  }

  # Shift up par 2 group numbers so they don't overlap with par 1 group
  # numbers.
  groups[[2]] <- groups[[2]] + max(groups[[1]])

  # Create a matrix of these values.
  mat <- matrix(no_group, nrow = length(partition), ncol = 3)
  mat[partition == 1, 1] <- groups[[1]]
  mat[partition == 2, 2] <- groups[[2]]
  mat[stem, 3] <- groups[[3]]

  # Set the values in the stem column to be larger than the values in the other
  # columns.
  mat[mat[,3] != -1, 3] <- mat[mat[,3] != -1, 3] + max(mat[, c(1,2)])

  # If there are no points in the stem, then retrieve the final group numbers.
  if(length(unique(mat[mat[,3] != no_group, 3])) == 0){
    mat <- mat[, 1:2, drop = FALSE]
    groups <- t(mat)
    groups <- groups[groups != no_group]
    groups <- as.numeric(factor(groups))
    return(groups)
  }

  # Structure incoming data.
  colnames(mat) <- c("pOne", "pTwo", "stem")
  mat <- as.data.frame(mat)

  # Link pOne to stem and update pTwo accordingly.
  ## Select pOne and stem.
  pOne <- mat[,c("pOne", "stem"), drop = FALSE]
  ## Determine the largest group number.
  maxGroup <- max(apply(mat, 2, max))
  ## Create a data.frame of the links to be made.
  pOne_dict <- pOne[
    (pOne[,"pOne"] != no_group) &
      (pOne[,"stem"] != no_group),
    , drop = FALSE
    ]
  ## Convert those values to factors.
  pOne_dict <- mutate_all(pOne_dict, as.factor)
  ## Perform linking.
  members <- components(
    graph.data.frame(
      pOne_dict,
      directed = FALSE
    ),
    mode = "weak"
  )$membership
  ## Convert mat to a matrix in order then update all group numbers accordingly.
  mat <- as.matrix(mat)
  logicalVector <- (mat != no_group) & (as.character(mat) %in% names(members))
  mat[logicalVector] <- members[as.character(mat[logicalVector])] + maxGroup
  mat <- as.data.frame(mat)

  # Link pOne to stem and update pTwo accordingly.
  ## Select pOne and stem.
  pTwo <- mat[,c("pTwo", "stem"), drop = FALSE]
  ## Determine the largest group number.
  maxGroup <- max(apply(mat, 2, max))
  ## Create a data.frame of the links to be made.
  pTwo_dict <- pTwo[
    (pTwo[,"pTwo"] != no_group) &
      (pTwo[,"stem"] != no_group),
    , drop = FALSE
    ]
  ## Convert those values to factors.
  pTwo_dict <- mutate_all(pTwo_dict, as.factor)
  ## Perform linking.
  members <- components(
    graph.data.frame(
      pTwo_dict,
      directed = FALSE
    ),
    mode = "weak"
  )$membership
  ## Convert mat to a matrix then update all group numbers accordingly.
  mat <- as.matrix(mat)
  logicalVector <- (mat != no_group) & (as.character(mat) %in% names(members))
  mat[logicalVector] <- members[as.character(mat[logicalVector])] + maxGroup

  # Select the pOne and pTwo columns
  mat <- mat[,c("pOne", "pTwo"), drop = FALSE]
  # Transpose the matrix and then remove all no_group values.
  mat <- t(mat)
  groups <- mat[mat != no_group]

  gc()
  as.numeric(factor(groups))
}



R/euclidean_linker.R
#' Group PALM/iPALM localizations based on their physical separation distance
#'
#' PALM/iPALM data results in a list of spatial coordinates for fluorophore
#' localizations. This function groups nearby localizations if they are within
#' the provided critical distance from each other.
#'
#' @title Return the group number for each localization.
#'
#' @param input A numeric matrix where each row is a localization and each
#' column is a spatial axis.
#' @param critDist The critical distance for which localizations nearer than
#' this distance are deemed part of the same group.
#' @param use_prog_bar TRUE/FALSE indicating whether a progress bar should be
#' used. This is only available when run_parallel is FALSE.
#' @param run_parallel TRUE/FALSE indicating whether operations should be
#' performed in parallel. This is only valid if partitioning is performed.
#' @param num_cores The number of cores to use if running in parallel.
#' @param partition_req The minimum number of points required to create a new
#' partition.
#' @param parallel_call_depth The number of levels of partitioning that should
#' be performed before terminating calls to run operations in parallel. The
#' number of threads opened when running in parallel is equal to
#' 2^(parallel_call_depth)*num_cores.
#' @param ... Additional parameters passed to euclidean_linker (i.e.
#' finding_blobs).
#'
#' @author Zach Colburn
#'
#' @examples
#' # Generate random data.
#' #set.seed(10)
#' #input <- as.matrix(data.frame(x=rnorm(10),y=rnorm(10)))
#'
#' # Perform linking.
#' #euclidean_linker(input, 0.4)
#'
#' @export
#'
#' @importFrom assertthat assert_that is.number is.flag
#' @importFrom parallel parLapply makeCluster detectCores
euclidean_linker <- function(
  input,
  critDist,
  use_prog_bar = TRUE,
  run_parallel = FALSE,
  num_cores = NULL,
  partition_req = 5000,
  parallel_call_depth = 3,
  ...
) {
  assert_that(class(input) == "matrix")
  assert_that(class(input[1]) %in% c("integer", "numeric"))
  assert_that(nrow(input) >= 1)
  assert_that(nrow(input) < 2147483646)# The C++ max limit for an int minus 1.
  assert_that(ncol(input) > 0)
  assert_that(is.number(critDist))
  assert_that(critDist > 0)
  assert_that(is.flag(use_prog_bar))
  assert_that(is.flag(run_parallel))
  assert_that(
    is.number(partition_req) &&
      (partition_req >= 100) &&
      (as.integer(partition_req) == partition_req)
  )
  assert_that(
    is.null(num_cores) ||
      is.number(num_cores)
  )
  if(run_parallel){use_prog_bar <- FALSE;}
  if(
    is.null(num_cores) && run_parallel
  ){
    num_cores <- detectCores()
  }else if(run_parallel){
    assert_that(
      (as.integer(num_cores) == num_cores) &&
        (num_cores >= 2) &&
        (num_cores <= detectCores())
    )
  }
  finding_blobs <- FALSE
  if("find_blobs" %in% names(match.call())){
    assert_that(is.flag(match.call()[["find_blobs"]]))
    finding_blobs <- match.call()[["find_blobs"]]
  }
  if(finding_blobs){
    min_gap <- 6
  }else{
    min_gap <- 6*critDist
  }

  # Return 1 if there is only a single input point.
  if(nrow(input) == 1){
    return(1)
  }

  # If partitioning:
  if(nrow(input) >= partition_req){
    # This value is used as an identifier of ungrouped points. It is used
    # in ".euclidean_linker_cpp" and ".perform_grouping" as well. The value
    # should not be changed.
    no_group <- -1

    groups <- .perform_partitioning(
      input,
      critDist = critDist,
      use_prog_bar = FALSE,
      run_parallel = run_parallel,
      num_cores = num_cores,
      partition_req = partition_req,
      parallel_call_depth = parallel_call_depth,
      min_gap = min_gap
    )
    output <- as.numeric(factor(groups))

    # Ensure memory from any parallelized threads is retrieved.
    gc()

    return(output)
  }

  # If not partitioning:
  output <- .perform_grouping(
    input,
    critDist,
    use_prog_bar = use_prog_bar
  )

  # Return the output.
  output
}

